### PR TITLE
Reset sessions

### PR DESCRIPTION
**Pull request by mmatera**
In this PR I added a method to ``MathicsSession`` that resets the ``Definitions`` object. This is useful to ensure that different pytests uses a "fresh" Definitions object.   

Also, I have improved the interface and the documentation of ``test.helper.check_evaluation`` to handle tests that compare warning messages and printed outs during the evaluation. 

<a href="https://gitpod.io/#https://github.com/Mathics3/mathics-core/pull/198"></a>

### CODE REVIEW

File: test/test_assignment.py [hunk 1]
```diff
@@ -83,17 +83,42 @@ def test_setdelayed_oneidentity():
             "{F[a, b], H[a, b], F[a, b]}",
             None,
         ),
+        (
```
Comment (mmatera) [inline test/test_assignment.py:86]: I added this "idiom" to tell to the test routine that a refresh is needed between tests. Maybe there is a more canonical way to to this, but for now it works. @rocky, @TiagoCavalcante, do you have an idea about how to do this better?

File: test/test_combinatorica.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
 # -*- coding: utf-8 -*-
-from .helper import evaluate, check_evaluation
+from .helper import evaluate, check_evaluation, reset_session
+
+
+def test_00_initialize():
```
Comment (mmatera) [inline test/test_combinatorica.py:5]: This is another thing that I am not sure about how to do. What I want here is that an initializing routine being run just before the tests. In the way that was implemented before, the `evaluate(...)` call was done before all the tests in `mathics-core/test/` be run, so a call to ``session.reset`` unloaded the required package.

Comment (rocky) [reply test/test_combinatorica.py:5]: > This is another thing that I am not sure about how to do. What I want here is that an initializing routine being run just before the tests. In the way that was implemented before, the `evaluate(...)` call was done before all the tests in `mathics-core/test/` be run, so a call to `session.reset` unloaded the required package.

https://stackoverflow.com/questions/22627659/run-code-before-and-after-each-test-in-py-test may help

**Comment by mmatera** [comment]
@rocky @TiagoCavalcante I think now this is ready for a final review.

**Pull request merged**